source_modelling/sources.py

```diff
@@ -217,8 +220,73 @@ class Plane:
          3            2
     """
 
-    # Bounds for plane are just the corners
-    bounds: np.ndarray
+    bounds: npt.NDArray[np.float32]
+
+    def _check_bounds_form_plane(self):
+        """Check that the bounds form a plane.
+
+        The checks performed are:
+            1. Two points with a minimum depth.
+            2. Two points with a maximum depth.
+            3. Four points total, spanning exactly a plane. Not a line or a point.
+
+        Raises
+        ------
+        ValueError
+            If the bounds do not form a plane.
+        """
+        top_mask = np.isclose(self.bounds[:, 2], self.bounds[:, 2].min())
+        top = self.bounds[top_mask]
+        bottom = self.bounds[~top_mask]
+        if (
+            np.linalg.matrix_rank(self.bounds) != 3
+            or len(top) != 2
+            or len(self.bounds) != 4
+            or not np.isclose(bottom[0, 2], bottom[1, 2])
+        ):
+            raise ValueError("Bounds do not form a plane.")
+
+    def __post_init__(self):
+        """Check that the plane bounds are consistent and in the correct order.
+
+        These checks are performed to ensure that the corners of the plane
+        actually form a plane, and the order of the points is in the order
+        required for the plane coordinates to make sense. The plane coordinates
+        require the first two points to be the top left and top right corners
+        (with respect to strike), and the last two points to be the bottom right
+        and bottom left corners (with respect to strike).
+
+        These checks are non-trivial which is why we implement them here instead
+        of requiring the caller to verify this themselves.
+
+        Raises
+        ------
+        ValueError
+            If the bounds do not form a plane.
+        """
+        self._check_bounds_form_plane()
+
+        top_mask = np.isclose(self.bounds[:, 2], self.bounds[:, 2].min())
+        top = self.bounds[top_mask]
+        bottom = self.bounds[~top_mask]
+        # We want to ensure that the bottom and top are pointing in roughly the
+        # same direction. To do this, we compare the dot product of the top and
+        # bottom vectors. If the dot product is positive then they are pointing
+        # in roughly the same direction, if it is negative then they are
+        # pointing in opposite directions.
+        if np.dot(top[1] - top[0], bottom[1] - bottom[0]) < 0:
+            bottom = bottom[::-1]
+        orientation = np.linalg.det(
+            np.array([top[1] - top[0], bottom[0] - top[0]])[:, :-1]
+        )
+
+        # If the orientation is not close to 0 and is negative, then dip
+        # direction is to the left of the strike direction, so we reverse
+        # the order of the top and bottom corners.
+        if not np.isclose(orientation, 0) and orientation < 0:
+            top = top[::-1]
+            bottom = bottom[::-1]
+        self.bounds = np.array([top[0], top[1], bottom[1], bottom[0]])
```

Comment:
As per comment above
```suggestion
        self.bounds = np.array([top[0], top[1], bottom[0], bottom[1]])
```